User input PR - needs testing

originalmartin97 opened the pull request:
not yet finished
needs some testing

jbakosi reviewed (changes requested):
Needs some tests. Please teach use how this code you are adding can be used by adding some unit tests. This will also prove that the code is correct and functions as intended.

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Leenkh3/Firefly/29#-:-OMLhGGF8axMo4TJ3M3B:b-mcmkbg)* status: all files reviewed, 4 unresolved discussions (waiting on @originalmartin97)

___
*[`user_input.c` line 7 at r1](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OMLgcPu40JAOFK97UMO:-OMLgcPu40JAOFK97UMP:b-siol4c) ([raw file](https://github.com/Leenkh3/Firefly/blob/8df427e320ba383f0740b7bc9735f28796cab52e/user_input.c#L7)):*
> ```C
> #include <lualib.h>
> 
> int main(int argc, char *argv[]) {
> ```

Let's create a tool for our customers! This file `user_input.c` has a main. The user will run the 'firefly' executable, which has an (s-far) empty main. I would convert this main into a function that performs the user input, call it from the main in firefly? No?
___
*[`user_input.c` line 93 at r1](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OMLgyQv8HMyxs4rb026:-OMLgyQv8HMyxs4rb027:bhwnlzr) ([raw file](https://github.com/Leenkh3/Firefly/blob/8df427e320ba383f0740b7bc9735f28796cab52e/user_input.c#L93)):*
> ```C
>     
>     // Clean up Lua
>     lua_close(L);
> ```

Nice work on the above lua-based parser. Now let's integrate it so we _create_ a new tool
___
*[`README.md` line 22 at r1](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OMLfe4F2eRc9Z-AAPMh:-OMLfe4F2eRc9Z-AAPMi:b6j0t3g) ([raw file](https://github.com/Leenkh3/Firefly/blob/8df427e320ba383f0740b7bc9735f28796cab52e/README.md#L22)):*
> ```Markdown
> #### To compile the C program with Lua support:
> 
> ```fish
> ```

You use [fish](https://fishshell.com/)? Nice!
___
*[`README.md` line 29 at r1](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OMLfnRoAgQlAhkvrK5p:-OMLfnRp5DeO4UZrfuRm:b56x8cp) ([raw file](https://github.com/Leenkh3/Firefly/blob/8df427e320ba383f0740b7bc9735f28796cab52e/README.md#L29)):*
> ```Markdown
> 
> ```
> gcc -std=c99 -o user_input user_input.c -llua -lm -ldl
> ```

Will we simply assume that the lua headers are available on the system? Or we want to be more user/developer friendly and detect instead and if not found download and install?


<!-- Sent from Reviewable.io -->

jbakosi commented:
Nice work! Let's add some new code that performs some tests. You can see multiple examples already from multiple team members already in the main branch, under, e.g., `src/test/`, `src/matrix/`, and `src/laplacian/`. And let's keep source code nice and organized under the `src/` directory! The root directory of the project should be very minimal, containing perhaps only a README and a CMakeLists.txt.

jbakosi commented:
Please look at the CI and fix the errors:
```sh
CMake Error at CMakeLists.txt:42 (add_executable):
  Cannot find source file:

    src/user_input/user_input.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm
  .ccm .cxxm .c++m .h .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90
  .f95 .f03 .hip .ispc


CMake Error at CMakeLists.txt:42 (add_executable):
  No SOURCES given to target: user_input


CMake Generate step failed.  Build files cannot be regenerated correctly.
Error: Process completed with exit code 1.
```

originalmartin97 commented:
Now it works properly only awaits for test creation for teaching purposes and making it to become a tool to integrate into main program

originalmartin97 commented:
made user_input as a function, now it can be integrated.

originalmartin97 commented:
now, it has some tests and also changed the documentations up a little bit, also added convinience factors like addition dependencies handling for the whole project (implemented within cmake as well)

jbakosi reviewed (changes requested):
Very nice work, overall, @originalmartin97!

Please see some relatively lengthy but minor comments. I really only have one "problem", which could well be related my system's setup, and as a result my user-input tests fail for some reason. I pasted the output in case that's helpful. You are the expert, maybe you cand spot the problem right away. Otherwise and in general, excellent work! I think is pretty close to merge, but would be good to address (find out) why that test fail.

> Reviewed 1 of 2 files at r2, 3 of 6 files at r3, 3 of 6 files at r4, 13 of 13 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Leenkh3/Firefly/29#-:-ONnGbmSDNjMT_eK7B1P:b-5fdskn)* status: all files reviewed, 9 unresolved discussions (waiting on @originalmartin97)

___
*[`README.md` line 94 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnBoP91ZdkhpgBWqzG:-ONnBoP91ZdkhpgBWqzH:b-knj0bx) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/README.md#L94)):*
> ```Markdown
> cmake -DLUA_INCLUDE_DIR=/path/to/lua/include -DLUA_LIBRARIES=/path/to/lua/lib/liblua.so ..
> ```
> ```

Nice job on updating the information in the README.
___
*[`README.md` line 113 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnBuiY6B_X5DupkJbB:-ONnBuiY6B_X5DupkJbC:bn95p78) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/README.md#L113)):*
> ```Markdown
> 
> 
> #### To compile the C program with Lua support:
> ```

Since we now will have nice instructions on building, this section is no longer of much use, no?
___
*[`scripts/install_dependencies.sh` line 83 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnD0Y955Lt4OJimhJj:-ONnD0Y955Lt4OJimhJk:b-1yd8xc) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/scripts/install_dependencies.sh#L83)):*
> ```Shell
> esac
> 
> echo "All dependencies installed successfully!"
> ```

This isn't a bad idea in general and helpful. However, if I just quickly want to try a project I randomly downloaded from github or I don't have root access, this will be less useful because I may not want some project I don't know much about executing commands and scripts with my root user. Another disadvantage is that it requires maintenance as operating systems change and get updated.
___
*[`test/CMakeLists.txt` line 81 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnE8v09M5m4Lk6ybZi:-ONnE8v09M5m4Lk6ybZj:b-tjkjl2) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/test/CMakeLists.txt#L81)):*
> ```CMake
> else()
>     message(STATUS "Skipping user input tests due to missing dependencies")
> endif()
> ```

Nice!

In general, however, be careful how much code you write. The more functionality one puts in, the more maintenance one needs to perform in the future.
___
*[`scripts/check_dependencies.sh` line 79 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnChAK7Jw1-KQnmfBy:-ONnChAK7Jw1-KQnmfBz:b-83sjoh) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/scripts/check_dependencies.sh#L79)):*
> ```Shell
> echo "  bash scripts/install_dependencies.sh"
> echo ""
> echo "For more detailed information, see the README.md"
> ```

Nice!
___
*[`test/user_input_tests/README.md` line 52 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnGMoI9ynfmjEnonDw:-ONnGMoI9ynfmjEnonDx:bbxei9v) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/test/user_input_tests/README.md#L52)):*
> ```Markdown
> ## Extending the Tests
> 
> To add more tests:
> ```

I like the extra section on how to add tests. Excellent docs!
___
*[`test/user_input_tests/advanced_user_input_test.cpp` line 248 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnExmx8Sf8Sc0n3GOB:-ONnExmy70RqA4Kithl9:b-i6cfvk) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/test/user_input_tests/advanced_user_input_test.cpp#L248)):*
> ```cpp
>     EXPECT_TRUE(error_output.find("Error running Lua function") != std::string::npos);
>     EXPECT_TRUE(error_output.find("Simulated error in get_user_input") != std::string::npos);
> }
> ```

_Advanced_ tests, indeed. Good job! This is a lot of code and requires some nontrivial Lua/C-ineraction expertise. _"With great freedom comes great responsibility."_ For risk associated with things like this, please see https://en.wikipedia.org/wiki/Bus_factor. However, you are doing a very good job in documenting what you are doing, so this would likely be okay in a larger project (if customizability is truly and decidedly required).
___
*[`CMakeLists.txt` line 25 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONn95BV0opJ1U4lIJER:-ONn95BV0opJ1U4lIJES:b-xvmv9o) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/CMakeLists.txt#L25)):*
> ```CMake
>         set(FIREFLY_SKIP_USER_INPUT_TESTS TRUE)
>     endif()
> endif()
> ```

Nice!

Do we _require_ lua? I think we do, independent of building tests, because we rely on this for user input, correct?

When I run this, it finds lua on my system, which is good, but the `ExternalPorject_Add()` called below will always build lua, which seems unnecessary in that case.

```sh
$ cd build && cmake ..
-- Found Lua: /usr/lib/x86_64-linux-gnu/liblua5.4.so;/usr/lib/x86_64-linux-gnu/libm.so (found version "5.4.4") 
-- Found Lua: /home/jbakosi/code/Firefly/src/user_input/LUA/include
$ make
...
[ 17%] Performing build step for 'lua'
Building Lua for Linux
-- Building Lua for platform:
-- Using generic posix target for
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
ar: `u' modifier ignored since `D' is the default (see `U')
[ 18%] Performing install step for 'lua'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
[ 19%] Completed 'lua'
[ 19%] Built target lua
[ 20%] Building C object CMakeFiles/numdiff.dir/src/ndiff/src/args.c.o
[ 21%] Building C object CMakeFiles/numdiff.dir/src/ndiff/src/constraint.c.o
...
```
___
*[`CMakeLists.txt` line 79 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONn9v7K-BKmcO66wCYA:-ONn9v7K-BKmcO66wCYB:b-ul7oqm) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/CMakeLists.txt#L79)):*
> ```CMake
>     COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_BINARY_DIR}/lua-install/include ${CMAKE_SOURCE_DIR}/src/user_input/LUA/include
>     COMMAND ${CMAKE_COMMAND} -E copy_directory ${PROJECT_BINARY_DIR}/lua-install/lib ${CMAKE_SOURCE_DIR}/src/user_input/LUA/lib
> )
> ```

Good job!
___
*[`CMakeLists.txt` line 83 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnA0Db6wZxLYNudo2c:-ONnA0Db6wZxLYNudo2d:baxyoba) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/CMakeLists.txt#L83)):*
> ```CMake
> # 1. Create the directories
> file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/src/user_input/LUA/include")
> file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/src/user_input/LUA/lib")
> ```

Creating directories in the `src/` dir is not considered good practice. In general, we only want to _write_ to the build directory. This keeps build outputs in one place and easy to work with.

If you need to find the lua includes and libs later within cmake you can use `${CMAKE_BUILD_DIR}` and specify a path relative to that.
___
*[`CMakeLists.txt` line 103 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnAYob4KgLCSinM_4u:-ONnAYocBtDzN6Mm7c1O:bzg2sql) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/CMakeLists.txt#L103)):*
> ```CMake
> add_dependencies(user_input lua)
> target_include_directories(user_input PRIVATE ${LUA_INCLUDE_DIR}) # This is now redundant but harmless
> target_link_libraries(user_input PRIVATE m) # 'm' for the math library
> ```

Otherwise, excellent job with the above!
___
*[`CMakeLists.txt` line 154 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnCBts-khzBdDiHcB_:-ONnCBts-khzBdDiHcBa:b-1j5zmb) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/CMakeLists.txt#L154)):*
> ```CMake
> if(FIREFLY_BUILD_TESTS)
>     add_subdirectory(test)
> endif()
> ```

When I run this on my linux, I get
```sh
$ ctest -j4
...
93% tests passed, 1 tests failed out of 14

Total Test time (real) =   0.04 sec

The following tests FAILED:
         14 - UserInputTests (Failed)
$ $ ctest -R UserInputTests -V
UpdateCTestConfiguration  from :/home/jbakosi/code/Firefly/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/jbakosi/code/Firefly/build/DartConfiguration.tcl
Test project /home/jbakosi/code/Firefly/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 14
    Start 14: UserInputTests

14: Test command: /home/jbakosi/code/Firefly/build/test/user_input_tests/user_input_tests
14: Working Directory: /home/jbakosi/code/Firefly/build/test/user_input_tests
14: Environment variables:
14:  FIREFLY_SCRIPT_PATH=/home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua
14: Test timeout computed to be: 10000000
14: Running main() from /home/jbakosi/code/Firefly/build/_deps/googletest-src/googletest/src/gtest_main.cc
14: [==========] Running 5 tests from 1 test suite.
14: [----------] Global test environment set-up.
14: [----------] 5 tests from UserInputTest
14: [ RUN      ] UserInputTest.ProcessHelp
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:83: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:84: Failure
14: Value of: output.find("Operation: help") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessHelp (0 ms)
14: [ RUN      ] UserInputTest.ProcessExport
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:102: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:103: Failure
14: Value of: output.find("Operation: export") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:104: Failure
14: Value of: output.find("Filetype: json") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:105: Failure
14: Value of: output.find("Filename: output.json") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessExport (0 ms)
14: [ RUN      ] UserInputTest.ProcessNoArguments
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:123: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:124: Failure
14: Value of: output.find("No arguments provided, starting interactive mode") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:125: Failure
14: Value of: output.find("Operation: load") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:126: Failure
14: Value of: output.find("Filetype: csv") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:127: Failure
14: Value of: output.find("Filename: input.csv") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessNoArguments (0 ms)
14: [ RUN      ] UserInputTest.ProcessExit
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:145: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:146: Failure
14: Value of: output.find("Operation: exit") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:147: Failure
14: Value of: output.find("Exiting program") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessExit (0 ms)
14: [ RUN      ] UserInputTest.ScriptNotFound
14: Starting program...
14: Loading Lua script from: /non/existent/path.lua
14: [       OK ] UserInputTest.ScriptNotFound (0 ms)
14: [----------] 5 tests from UserInputTest (1 ms total)
14:
14: [----------] Global test environment tear-down
14: [==========] 5 tests from 1 test suite ran. (1 ms total)
14: [  PASSED  ] 1 test.
14: [  FAILED  ] 4 tests, listed below:
14: [  FAILED  ] UserInputTest.ProcessHelp
14: [  FAILED  ] UserInputTest.ProcessExport
14: [  FAILED  ] UserInputTest.ProcessNoArguments
14: [  FAILED  ] UserInputTest.ProcessExit
14:
14:  4 FAILED TESTS
1/1 Test #14: UserInputTests ...................***Failed    0.00 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.01 sec

The following tests FAILED:
         14 - UserInputTests (Failed)
Errors while running CTest
Output from these tests are in: /home/jbakosi/code/Firefly/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```


<!-- Sent from Reviewable.io -->

originalmartin97 commented:
@jbakosi 
> > ```cmake
> > # 1. Create the directories
> > file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/src/user_input/LUA/include")
> > file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/src/user_input/LUA/lib")
> > ```
> 
> Creating directories in the `src/` dir is not considered good practice. In general, we only want to _write_ to the build directory. This keeps build outputs in one place and easy to work with.
> 
> If you need to find the lua includes and libs later within cmake you can use `${CMAKE_BUILD_DIR}` and specify a path relative to that.
>

Also, I did not seem to find this part within any of my CMakeLists.txt files. I suppose I already fixed it earlier.

jbakosi reviewed (changes requested):
> To be honest, I could not understand completely why this Failed for you

I also don't understand. But I also do not understand why the CI is not running the user input tests:
https://github.com/Leenkh3/Firefly/actions/runs/14578791155/job/40890709434?pr=29

On the other hand, it now runs when I try it on my linux box:
```sh
$ ctest -R UserInputTests -V
UpdateCTestConfiguration  from :/home/jbakosi/code/Firefly/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/jbakosi/code/Firefly/build/DartConfiguration.tcl
Test project /home/jbakosi/code/Firefly/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 16
    Start 16: UserInputTests

16: Test command: /home/jbakosi/code/Firefly/build/src/test/user_input_tests/user_input_tests
16: Working Directory: /home/jbakosi/code/Firefly/build/src/test/user_input_tests
16: Environment variables: 
16:  FIREFLY_SCRIPT_PATH=/home/jbakosi/code/Firefly/build/src/test/user_input_tests/test_input_handler.lua
16: Test timeout computed to be: 10000000
16: Running main() from /home/jbakosi/code/Firefly/build/_deps/googletest-src/googletest/src/gtest_main.cc
16: [==========] Running 5 tests from 1 test suite.
16: [----------] Global test environment set-up.
16: [----------] 5 tests from UserInputTest
16: [ RUN      ] UserInputTest.ProcessHelp
16: [       OK ] UserInputTest.ProcessHelp (0 ms)
16: [ RUN      ] UserInputTest.ProcessExport
16: [       OK ] UserInputTest.ProcessExport (0 ms)
16: [ RUN      ] UserInputTest.ProcessNoArguments
16: [       OK ] UserInputTest.ProcessNoArguments (0 ms)
16: [ RUN      ] UserInputTest.ProcessExit
16: [       OK ] UserInputTest.ProcessExit (0 ms)
16: [ RUN      ] UserInputTest.ScriptNotFound
16: Starting program...
16: Loading Lua script from: /non/existent/path.lua
16: [       OK ] UserInputTest.ScriptNotFound (0 ms)
16: [----------] 5 tests from UserInputTest (1 ms total)
16: 
16: [----------] Global test environment tear-down
16: [==========] 5 tests from 1 test suite ran. (1 ms total)
16: [  PASSED  ] 5 tests.
1/1 Test #16: UserInputTests ...................   Passed    0.00 sec

The following tests passed:
        UserInputTests

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.01 sec
```

> Reviewed 22 of 22 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Leenkh3/Firefly/29#-:-OOWfA7JBRPIdZfP-isE:b2bmf9t)* status: all files reviewed, 13 unresolved discussions (waiting on @originalmartin97)

___
*[`CMakeLists.txt` line 154 at r5](https://reviewable.io/reviews/Leenkh3/Firefly/29#-ONnCBts-khzBdDiHcB_:-OOWhmj93JMIbiqwtDqh:b-g5owk2) ([raw file](https://github.com/Leenkh3/Firefly/blob/a9e8d762e6da0b5820306f1bbf307f26e88f963d/CMakeLists.txt#L154)):*
<details><summary><i>Previously, jbakosi (Jozsef Bakosi) wrote…</i></summary><blockquote>

When I run this on my linux, I get
```sh
$ ctest -j4
...
93% tests passed, 1 tests failed out of 14

Total Test time (real) =   0.04 sec

The following tests FAILED:
         14 - UserInputTests (Failed)
$ $ ctest -R UserInputTests -V
UpdateCTestConfiguration  from :/home/jbakosi/code/Firefly/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/jbakosi/code/Firefly/build/DartConfiguration.tcl
Test project /home/jbakosi/code/Firefly/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 14
    Start 14: UserInputTests

14: Test command: /home/jbakosi/code/Firefly/build/test/user_input_tests/user_input_tests
14: Working Directory: /home/jbakosi/code/Firefly/build/test/user_input_tests
14: Environment variables:
14:  FIREFLY_SCRIPT_PATH=/home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua
14: Test timeout computed to be: 10000000
14: Running main() from /home/jbakosi/code/Firefly/build/_deps/googletest-src/googletest/src/gtest_main.cc
14: [==========] Running 5 tests from 1 test suite.
14: [----------] Global test environment set-up.
14: [----------] 5 tests from UserInputTest
14: [ RUN      ] UserInputTest.ProcessHelp
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:83: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:84: Failure
14: Value of: output.find("Operation: help") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessHelp (0 ms)
14: [ RUN      ] UserInputTest.ProcessExport
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:102: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:103: Failure
14: Value of: output.find("Operation: export") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:104: Failure
14: Value of: output.find("Filetype: json") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:105: Failure
14: Value of: output.find("Filename: output.json") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessExport (0 ms)
14: [ RUN      ] UserInputTest.ProcessNoArguments
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:123: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:124: Failure
14: Value of: output.find("No arguments provided, starting interactive mode") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:125: Failure
14: Value of: output.find("Operation: load") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:126: Failure
14: Value of: output.find("Filetype: csv") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:127: Failure
14: Value of: output.find("Filename: input.csv") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessNoArguments (0 ms)
14: [ RUN      ] UserInputTest.ProcessExit
14: Error: input_handler.lua not found at /home/jbakosi/code/Firefly/test/user_input_tests/test_input_handler.lua!
14: Create it from template: cp ./src/user_input/input_handler.template.lua ./src/user_input/input_handler.lua
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:145: Failure
14: Expected equality of these values:
14:   0
14:   result
14:     Which is: 1
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:146: Failure
14: Value of: output.find("Operation: exit") != std::string::npos
14:   Actual: false
14: Expected: true
14: /home/jbakosi/code/Firefly/test/user_input_tests/user_input_test.cpp:147: Failure
14: Value of: output.find("Exiting program") != std::string::npos
14:   Actual: false
14: Expected: true
14: [  FAILED  ] UserInputTest.ProcessExit (0 ms)
14: [ RUN      ] UserInputTest.ScriptNotFound
14: Starting program...
14: Loading Lua script from: /non/existent/path.lua
14: [       OK ] UserInputTest.ScriptNotFound (0 ms)
14: [----------] 5 tests from UserInputTest (1 ms total)
14:
14: [----------] Global test environment tear-down
14: [==========] 5 tests from 1 test suite ran. (1 ms total)
14: [  PASSED  ] 1 test.
14: [  FAILED  ] 4 tests, listed below:
14: [  FAILED  ] UserInputTest.ProcessHelp
14: [  FAILED  ] UserInputTest.ProcessExport
14: [  FAILED  ] UserInputTest.ProcessNoArguments
14: [  FAILED  ] UserInputTest.ProcessExit
14:
14:  4 FAILED TESTS
1/1 Test #14: UserInputTests ...................***Failed    0.00 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.01 sec

The following tests FAILED:
         14 - UserInputTests (Failed)
Errors while running CTest
Output from these tests are in: /home/jbakosi/code/Firefly/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```
</blockquote></details>

This seems to be okay. See other comment.
___
*[`src/test/CMakeLists.txt` line 13 at r6](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OOWgWB8D2f7sTAruztv:-OOWgWB8D2f7sTAruztw:b-k0habk) ([raw file](https://github.com/Leenkh3/Firefly/blob/f7e7afb0821ebb77dc34ef67abb99b9efcc399fe/src/test/CMakeLists.txt#L13)):*
> ```CMake
>     message(STATUS "  Arch Linux:    sudo pacman -S lua")
>     message(STATUS "  macOS:         brew install lua")
>     set(SKIP_USER_INPUT_TESTS TRUE)
> ```

I would argue that user input is a pretty important part of this application, so we should make sure that works. Hence, I would _require_ the user input tests and not skip it if lua is not found. (If lua is not found, we need to build it fail hard lua cannot be made available),
___
*[`src/test/user_input_tests/test.bak/CMakeLists.txt` line 1 at r6](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OOWiFgc9IQtIeM7RTyL:-OOWiFgc9IQtIeM7RTyM:b-3v5l56) ([raw file](https://github.com/Leenkh3/Firefly/blob/f7e7afb0821ebb77dc34ef67abb99b9efcc399fe/src/test/user_input_tests/test.bak/CMakeLists.txt#L1)):*
> ```CMake
> cmake_minimum_required(VERSION 3.10)
> ```

This is now renamed/moved from `test/CMakeLists.txt` to `src/test/user_input_tests/test.bak/CMakeLists.txt`. Do we still need to keep it?
___
*[`src/user_input/logs/2025-04-21.md` line 31 at r6](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OOWiXGrAAZ26yARSNrN:-OOWiXGrAAZ26yARSNrO:b90qy60) ([raw file](https://github.com/Leenkh3/Firefly/blob/f7e7afb0821ebb77dc34ef67abb99b9efcc399fe/src/user_input/logs/2025-04-21.md#L31)):*
> ```Markdown
> 
> ## Next Steps
> - resolve conflicts with branch main
> ```

Good work taking notes! (I do something similar.) But likely not necessary publish it by committing it to the project repository.
___
*[`src/test/user_input_tests/test.bak/user_input_tests/user_input_test.cpp` line 1 at r6](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OOWiRZR12oDXZ5U8-n0:-OOWiRZR12oDXZ5U8-n1:b3edwvp) ([raw file](https://github.com/Leenkh3/Firefly/blob/f7e7afb0821ebb77dc34ef67abb99b9efcc399fe/src/test/user_input_tests/test.bak/user_input_tests/user_input_test.cpp#L1)):*
> ```cpp
> #include <gtest/gtest.h>
> ```

Same here and a few other files above.
___
*[`firefly.cpp` line 6 at r6](https://reviewable.io/reviews/Leenkh3/Firefly/29#-OOWir7Y71HWgMTMcAsW:-OOWir7Y71HWgMTMcAsX:b-3k5gxx) ([raw file](https://github.com/Leenkh3/Firefly/blob/f7e7afb0821ebb77dc34ef67abb99b9efcc399fe/firefly.cpp#L6)):*
> ```cpp
> {
>     std::cout << "Firefly Project - Main Execution" << std::endl;
>     return 0;
> ```

Our main executable still only does a printout.


<!-- Sent from Reviewable.io -->
